File: dotnet-adapters-shortest-paths-core/test/java/com/programmerare/shortestpaths/core/impl/generics/PathGenericsImplTest.java
/*
* Copyright (c) Tomas Johansson , http://www.programmerare.com
* The code is made available under the terms of the MIT License.
* https://github.com/TomasJohansson/adapters-shortest-paths/blob/master/adapters-shortest-paths-core/License.txt
*/
package com.programmerare.shortestpaths.core.impl.generics;

import static com.programmerare.shortestpaths.core.impl.PathImpl.createPath;
import static com.programmerare.shortestpaths.core.impl.EdgeImpl.createEdge;
import static com.programmerare.shortestpaths.core.impl.VertexImpl.createVertex;
import static com.programmerare.shortestpaths.core.impl.WeightImpl.SMALL_DELTA_VALUE_FOR_WEIGHT_COMPARISONS;
import static com.programmerare.shortestpaths.core.impl.WeightImpl.createWeight;
import static com.programmerare.shortestpaths.core.impl.generics.PathGenericsImpl.createPathGenerics;
import static org.junit.Assert.assertEquals;

import java.util.Arrays;
import java.util.List;

import org.junit.Before;
import org.junit.Test;

import com.programmerare.shortestpaths.core.api.Edge;
import com.programmerare.shortestpaths.core.api.Path;
import com.programmerare.shortestpaths.core.api.generics.EdgeGenerics;

/**
 * @author Tomas Johansson
 */
public class PathGenericsImplTest {

	private Edge edgeAB3;
	private Edge edgeBC5;
	private Edge edgeCD7;
	private String firstVertex, secondVertex, thirdVertex, fourthVertex;
	private double weightFirstEdge, weightSecondEdge, weightThirdEdge, totalWeight;
	private Path path; 
			
	@Before
	public void setUp() throws Exception {
		firstVertex = "A";
		secondVertex = "B";
		thirdVertex = "C";
		fourthVertex = "D";
	
		weightFirstEdge = 3;
		weightSecondEdge = 5;
		weightThirdEdge = 7;
		totalWeight = weightFirstEdge + weightSecondEdge + weightThirdEdge;
		
		edgeAB3 = createEdge(createVertex(firstVertex), createVertex(secondVertex), createWeight(weightFirstEdge));
		edgeBC5 = createEdge(createVertex(secondVertex), createVertex(thirdVertex), createWeight(weightSecondEdge));
		edgeCD7 = createEdge(createVertex(thirdVertex), createVertex(fourthVertex), createWeight(weightThirdEdge));
		
		path = createPath(createWeight(totalWeight), Arrays.asList(edgeAB3, edgeBC5, edgeCD7));
	}

	@Test
	public void testGetTotalWeightForPath() {
		assertEquals(totalWeight, path.getTotalWeightForPath().getWeightValue(), SMALL_DELTA_VALUE_FOR_WEIGHT_COMPARISONS);
	}

	@Test
	public void testGetEdgesForPath() {
		List<? extends EdgeGenerics> edgesForPath = path.getEdgesForPath();
		assertEquals(3, edgesForPath.size());
		assertEquals(edgeAB3, edgesForPath.get(0));
		assertEquals(edgeBC5, edgesForPath.get(1));
		assertEquals(edgeCD7, edgesForPath.get(2));
	}

	@Test(expected = RuntimeException.class) 
	public void testExceptionIsThrownIfVerticesIsNotMatching() { 
		createPathGenerics(
			createWeight(15d),  
			Arrays.asList(
				createEdge(createVertex("A"), createVertex("B"), createWeight(3d)),
				createEdge(createVertex("B"), createVertex("C"), createWeight(5d)),
				 // Note that "X" should be "C" below, which is the reason for expected exceotion
				createEdge(createVertex("X"), createVertex("D"), createWeight(7d))
			),
			false,
			true // tell creation method to throw exception if not all vertices are matching 				
		);
	}
	
	@Test(expected = RuntimeException.class) 
	public void testExceptionIsTotalWeightIsNotMatching() {
		createPathGenerics(
			createWeight(16), // SHOULD be 15 ( 3 + 5 + 7 ) and therefore an exception should be thrown 
			Arrays.asList(
				createEdge(createVertex("A"), createVertex("B"), createWeight(3d)),
				createEdge(createVertex("B"), createVertex("C"), createWeight(5d)),
				createEdge(createVertex("C"), createVertex("D"), createWeight(7d))
			),
			true,  // tell creation method to throw exception if sum is not matching
			false 				
		);
	}	
}

File: Programmerare.ShortestPaths.Test/test/Programmerare.ShortestPaths/Core/Impl/generics/PathGenericsImplTest.cs
/*
* Copyright (c) Tomas Johansson , http://www.programmerare.com
* The code in the "core" project is licensed with MIT.
* Other projects within this Visual Studio solution may be released with other licenses e.g. Apache.
* Please find more information in the files "License.txt" and "NOTICE.txt" 
* in the project root directory and/or in the solution root directory.
* It should also be possible to find more license information at this URL:
* https://github.com/TomasJohansson/adapters-shortest-paths-dotnet/
*/

using Programmerare.ShortestPaths.Core.Api;
using NUnit.Framework;
using static NUnit.Framework.Assert;
using static Programmerare.ShortestPaths.Core.Impl.WeightImpl; // SMALL_DELTA_VALUE_FOR_WEIGHT_COMPARISONS
using static Programmerare.ShortestPaths.Core.Impl.EdgeImpl; // createEdge
using static Programmerare.ShortestPaths.Core.Impl.VertexImpl; // createVertex
using System.Collections.Generic;
using static Programmerare.ShortestPaths.Core.Impl.PathImpl; // createPath
using System;

namespace Programmerare.ShortestPaths.Core.Impl.Generics
{
    /**
     * @author Tomas Johansson
     */
    [TestFixture]
    public class PathGenericsImplTest {

	    private Edge edgeAB3;
	    private Edge edgeBC5;
	    private Edge edgeCD7;
	    private string firstVertex, secondVertex, thirdVertex, fourthVertex;
	    private double weightFirstEdge, weightSecondEdge, weightThirdEdge, totalWeight;
	    private Path path; 
			
	    [SetUp]
	    public void setUp()  {
		    firstVertex = "A";
		    secondVertex = "B";
		    thirdVertex = "C";
		    fourthVertex = "D";
	
		    weightFirstEdge = 3;
		    weightSecondEdge = 5;
		    weightThirdEdge = 7;
		    totalWeight = weightFirstEdge + weightSecondEdge + weightThirdEdge;
		
		    edgeAB3 = CreateEdge(CreateVertex(firstVertex), CreateVertex(secondVertex), CreateWeight(weightFirstEdge));
		    edgeBC5 = CreateEdge(CreateVertex(secondVertex), CreateVertex(thirdVertex), CreateWeight(weightSecondEdge));
		    edgeCD7 = CreateEdge(CreateVertex(thirdVertex), CreateVertex(fourthVertex), CreateWeight(weightThirdEdge));
		
		    path = CreatePath(CreateWeight(totalWeight), new List<Edge> {edgeAB3, edgeBC5, edgeCD7 });
	    }

	    [Test]
	    public void testGetTotalWeightForPath() {
		    AreEqual(totalWeight, path.TotalWeightForPath.WeightValue, SMALL_DELTA_VALUE_FOR_WEIGHT_COMPARISONS);
	    }

	    [Test]
	    public void testGetEdgesForPath() {
		    var edgesForPath = path.EdgesForPath;
		    AreEqual(3, edgesForPath.Count);
		    AreEqual(edgeAB3, edgesForPath[0]);
		    AreEqual(edgeBC5, edgesForPath[1]);
		    AreEqual(edgeCD7, edgesForPath[2]);
	    }

	    [Test]//@Test(expected = RuntimeException.class) 
	    public void testExceptionIsThrownIfVerticesIsNotMatching() { 
            var exceptionThrown = Assert.Throws<Exception>(() => {
                PathGenericsImpl<Edge, Vertex, Weight>.CreatePathGenerics<Edge, Vertex, Weight>(
                    CreateWeight(15d),
                    new List<Edge>{
                        CreateEdge(CreateVertex("A"), CreateVertex("B"), CreateWeight(3d)),
                        CreateEdge(CreateVertex("B"), CreateVertex("C"), CreateWeight(5d)),
                        // Note that "X" should be "C" below, which is the reason for expected exceotion
                        CreateEdge(CreateVertex("X"), CreateVertex("D"), CreateWeight(7d))
                    },
                    false,
                    true // tell creation method to throw exception if not all vertices are matching
                );
            });
            IsNotNull(exceptionThrown);
        }
    
        [Test]
	    public void testExceptionIsTotalWeightIsNotMatching() {
            var exceptionThrown = Assert.Throws<Exception>(() => {
                PathGenericsImpl<Edge, Vertex, Weight>.CreatePathGenerics<Edge, Vertex, Weight>(
                    CreateWeight(16), // SHOULD be 15 ( 3 + 5 + 7 ) and therefore an exception should be thrown 
                    new List<Edge>{
                        CreateEdge(CreateVertex("A"), CreateVertex("B"), CreateWeight(3d)),
                        CreateEdge(CreateVertex("B"), CreateVertex("C"), CreateWeight(5d)),
                        CreateEdge(CreateVertex("C"), CreateVertex("D"), CreateWeight(7d))
                    },
                    true,  // tell creation method to throw exception if sum is not matching
                    false
                );
            });
            IsNotNull(exceptionThrown);
        }	
    }
}